tripy/tripy/frontend/trace/ops/shape.py

```diff
@@ -63,4 +63,12 @@ def shape(self: "tripy.Tensor") -> List["tripy.DimensionSize"]:
         assert shape == [8, 2]
     """
 
-    return [GetDimensionSize.build([self], dim=index, always_cast_to_dimension_size=True) for index in range(self.rank)]
+    from tripy.frontend.dimension_size import DimensionSize
+
+    # If the shape is statically known, we do not need to insert any operator calls and we can memoize.
+    if self.shape_memo is None and all(dim >= 0 for dim in self.trace_tensor.shape):
+        self.shape_memo = [DimensionSize(dim) for dim in self.trace_tensor.shape]
+
+    return self.shape_memo or [
+        GetDimensionSize.build([self], dim=index, always_cast_to_dimension_size=True) for index in range(self.rank)
```

Comment:
Memoizing the `GetDimensionSize` ops led to an MLIR error upon repeated calls, and I'm not entirely sure why.

The case that fails (doc for repeat):
```python
inp = tp.reshape(tp.arange(4, dtype=tp.int32), (2, 2))
out0 = tp.repeat(inp, 2, dim=0)
out1 = tp.repeat(inp, 2, dim=1)

np_inp = np.from_dlpack(tp.copy(inp, device=tp.device("cpu"))) # doc: omit
ref_out0 = np.repeat(np_inp, 2, 0)
assert np.array_equal(ref_out0, np.from_dlpack(tp.copy(out0, device=tp.device("cpu")))) # <- succeeds

ref_out1 = np.repeat(np_inp, 2, 1)
assert np.array_equal(ref_out1, np.from_dlpack(tp.copy(out1, device=tp.device("cpu")))) # <- fails
```

Error:
```
    (t347)error: number of output elements (8) doesn't match expected number of elements (4)

    This error occured while trying to compile the following FlatIR expression:
          |
          | t347: [rank=(3), shape=((-1, -1, -1)), dtype=(int32), loc=(gpu:0)] = DynamicReshapeOp(t305, t346)
          | 

    Note: This originated from the following expression:

    --> /tripy/docs/../tripy/frontend/ops/unsqueeze.py:56 in unsqueeze()
          |
       56 |     return reshape(input, result_shape)
          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    --> /tripy/docs/../tripy/frontend/ops/repeat.py:92 in repeat()
          |
       92 |     out = unsqueeze(input, dim + 1)
          |           ^^^^^^^^^^^^^^^^^^^^^^^^^ --- required from here
```

The unsqueeze gets the shape of the input and computes a new shape as an argument to `reshape`. I will see if I can figure out how that results in an error when lowering to MLIR.